Topics/Image Processing/Convolution/Convolution.pde

```diff
@@ -1,31 +1,77 @@
 /**
  * Convolution
- * by Daniel Shiffman.  
- * 
- * Applies a convolution matrix to a portion of an image. Move mouse to 
- * apply filter to different parts of the image.
+ * by Daniel Shiffman.
+ *
+ * Applies a convolution matrix to a portion of an image. Move mouse to
+ * apply filter to different parts of the image. Click mouse to cycle
+ * through different effects (kernels).
  */
 
 PImage img;
+int effect = 0;
 int w = 120;
 
-// It's possible to convolve the image with many different 
-// matrices to produce different effects. This is a high-pass 
-// filter; it accentuates the edges. 
-float[][] matrix = { { -1, -1, -1 },
-                     { -1,  9, -1 },
-                     { -1, -1, -1 } }; 
+// It's possible to convolve the image with many different
+// matrices to produce different effects.
+String[] effect_names = {
+  "Identity (no change)",
+  "Darken",
+  "Lighten",
+  "Sharpen",
+  "Sharpen More",
+  "Box Blur",
+  "Edge Detect",
+  "Emboss"
+};
+
+float[][][] matrix = { { { 0, 0, 0 },    // Identity
+                         { 0, 1, 0 },
+                         { 0, 0, 0 } },
+                       { { 0,   0, 0 },  // Darken
+                         { 0, 0.5, 0 },
+                         { 0,   0, 0 } },
+                       { { 0, 0, 0 },    // Lighten
+                         { 0, 2, 0 },
+                         { 0, 0, 0 } },
+                       { {  0, -1,  0 }, // Sharpen (high-pass)
+                         { -1,  5, -1 },
+                         {  0, -1,  0 } },
+                       { { -1, -1, -1 }, // Sharpen (stronger)
+                         { -1,  9, -1 },
+                         { -1, -1, -1 } },
+                       { { 1.0/9.0, 1.0/9.0, 1.0/9.0 }, // Box Blur (low-pass)
+                         { 1.0/9.0, 1.0/9.0, 1.0/9.0 },
+                         { 1.0/9.0, 1.0/9.0, 1.0/9.0 } },
+                       { { 0,  1, 0 },   // Edge Detect
+                         { 1, -4, 1 },
+                         { 0,  1, 0 } },
+                       { { -2, -1, 0 },  // Emboss
+                         { -1,  1, 1 },
+                         {  0,  1, 2 } } };
 
 void setup() {
   size(640, 360);
-  img = loadImage("moon-wide.jpg"); 
+  img = loadImage("moon-wide.jpg");
+
+  noLoop();
+}
+
+void mouseClicked() {
+  effect++;
+  if (effect >= effect_names.length) effect = 0;
+
+  redraw();
+}
+
+void mouseMoved() {
+  redraw();
```

Comment:
This overrides the mouse click meaning that you can't change mode while moving the mouse

Reply:
changed `mouseClicked` to `mousePressed`, added a `mouseDragged` handler too